rpsl-tools/main.go

```diff
@@ -0,0 +1,186 @@
+// with the rpsl-parser library and generates loadable data for indexed
+// queries of the irr content.
+package main
+
+import (
+	"compress/gzip"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	glog "github.com/golang/glog"
+	rpsl "github.com/manrs-tools/contrib/rpsl-parser"
+)
+
+var (
+	threads = flag.Int("threads", 4, "Max threads to use in parsing db files.")
+)
+
+// Define a slice of filenames to get as a single flag on startup.
+type fileSlice []string
+
+var rpslFiles fileSlice
+
+// Stirng and Set are required for the variadic flag: rpslFiles.
+func (f *fileSlice) String() string {
+	return fmt.Sprintf("%v", *f)
+}
+
+func (f *fileSlice) Set(value string) error {
+	for _, fn := range strings.Split(value, ",") {
+		*f = append(*f, fn)
+	}
+	return nil
+}
+
+// getReader parses the filename and either returns a gzip reader or regular reader.
+func getReader(fn string) (io.Reader, error) {
+	fd, err := os.Open(fn)
+	if err != nil {
+		fmt.Printf("Failed to open file(%v): %v\n", fn, err)
+		return nil, err
+	}
+	if strings.HasSuffix(fn, ".gz") {
+		f, err := gzip.NewReader(fd)
+		if err != nil {
+			fmt.Printf("Failed to read gzip'd file(%v): %v\n", fn, err)
+			return nil, err
+		}
+		return f, nil
+	}
+	return fd, nil
+}
+
+// parseFile parses a single file at a time into records sent to the
+// records channel (rc). Files arrive as names on the input channel (ic).
+func parseFile(ic <-chan string, rc chan<- *rpsl.Record, ec chan<- bool) {
+	for fn := range ic {
+		var rdr *rpsl.Reader
+
+		reader, err := getReader(fn)
+		if err != nil {
+			glog.Infof("failed to read file(%v): %v\n", fn, err)
+			continue
+		}
+		rdr = rpsl.NewReader(reader)
+
+		// Read all leading comments and whitespace.
+		err = rdr.ConsumeComment()
+		if err != nil {
+			glog.Infof("Failed reading file(%v): %v\n", fn, err)
+			continue
+		}
+		err = rdr.ConsumeLeadingWS()
+		if err != nil {
+			glog.Infof("Failed reading file(%v): %v\n", fn, err)
+			continue
+		}
+
+		// The file must start with a letter, all IRR records start with a letter character.
+		r := rdr.Peek()
+		if !rpsl.IsLetter(r) {
+			glog.Infof("The first character read(%v) is not a letter, file unparsable.\n", string(r))
+			// Add 2 more chars so finding the problem is more possible.
+			r, _, _ := rdr.Read()
+			glog.Infof("Next char: %v\n", string(r))
+			r, _, _ = rdr.Read()
+			glog.Infof("Next char: %v\n", string(r))
+			continue
+		}
+
+		// Parse the file, sending results back up the channel (rc).
+		rpsl.Parse(rdr, rc)
+
+		// When done parsing, send a file-done message on the end channel (ec).
+		ec <- true
+	}
+}
+
+// Verify that there files requested exist, open each in a goroutine and feed
+// those to the rpsl-parser library, returning each record to a channel for
+// disposition in the final data structure to be loaded into a DB.
+func main() {
+	flag.Var(&rpslFiles, "rpslFiles", "Files to parse, irr/rpsl content(csv).")
```

Comment:
does this mean the file contents should be csv, or that the list of files are comma separated?  presumably the latter?

Reply:
list of files is csv.
the actual file content is the 'standard' RPSL backup content.

Reply:
Clarified in the flag guidance.